gh-107801: Improve io.*.seek docs and docstrings (#107899)
python/cpython: Doc/library/io.rst

   :func:`os.stat`) if possible.

> .. data:: SEEK_SET

Reviewer: These constants are defined in `os` and `io` modules. Instead of duplicating the description, I think that it is better to document them in one place and refer to them from other module.

Author: They are exposed API in both modules, so in my opinion they should be documented in both `io` and `os` docs. I think `.. seealso::` for cross linking is fine.

          SEEK_CUR
          SEEK_END

>    Parameters to the :meth:`~IOBase.seek` method,
>    used to specify the relative position
>    Their values are 0, 1, and 2, respectively.

Reviewer 2: Perhaps a bit verbose, but I think it might be useful to spell out what each constant means. Also this allows putting the values in-line.

```suggestion
   Parameters to the :meth:`~IOBase.seek` method,
   used to specify the relative position to seek from.

   :data:`!SEEK_SET`
     Seek from the start of the stream
     (Value: ``0``)
   :data:`!SEEK_CUR`
     Seek from the current stream position
     (Value: ``1``)
   :data:`!SEEK_END`
     Seek from the end of the stream
     (Value: ``2``)

```

Author: Thanks; applied but not pushed.

   .. seealso::

>       :data:`os.SEEK_SET`, :data:`os.SEEK_CUR`, and :data:`os.SEEK_END`.

Reviewer 2: It seems the ``os`` constants were added in 2.5 for the ``os.lseek`` function (and ``io``'s in 2.7 & 3.1 (#48822)). Given the values are the same there's little difference, but should we provide which module's constants you're 'meant' to use in which cases?

Edit: Note that ``fnctl.lockf``, ``mmap.seek`` and ``sqlite3.Blob.seek`` use the ``os`` constants.

Reviewer: As `os` constants were added earlier, I think we can just refer to them from here.

Author: I have no opinion on which constants to use, but my gut reaction would be to use the constants from `io` for `io` stuff.

   .. method:: seek(offset, whence=SEEK_SET, /)

      Change the stream position to the given byte *offset*,

>       and return the new absolute position as an integer.

Reviewer: Returning the new position is secondary and was already documented at the end of the method description. Is it so important to move it to the beginning?

Author: If a function or method returns a value, surely it makes sense to start the text with that information, together with a succinct description of what the function/method does.

      *offset* is interpreted relative to the position indicated by *whence*.
      Valid values for *whence* are:

>       * :data:`SEEK_SET` -- seek from the start of the stream (the default);

Reviewer: Please keep the literal values. They are standard and widely used. Removing the literal values will make the reader confused when they see `file.seek(0, 2)` in someone's other code.

Reviewer: The text above contains: "*offset* is interpreted relative to the position indicated by *whence*." The text in this paragraph describes that position. So if you consider it as a continuation of the former phrase, "seek from" may be not needed.

Author: > Please keep the literal values. They are standard and widely used. Removing the literal values will make the reader confused when they see `file.seek(0, 2)` in someone's other code.

Good call. I agree.

Reviewer 3: I would also keep the literal values.

        *offset* should be zero or positive

>       * :data:`SEEK_CUR` -- seek from current stream position;

Reviewer 2: ```suggestion
      * :data:`SEEK_CUR` -- seek from the current stream position;
```

        *offset* may be negative
      * :data:`SEEK_END` -- seek from the end of the stream;

>         *offset* is usually negative

Reviewer: ```suggestion
        *offset* is usually zero or negative
```

> 

Reviewer 2: Given the pitfalls of seeking with files in text-mode, perhaps ``IOBase.seek`` should link to ``TextIOBase.seek``, which has more specific information on the limitations here:

```suggestion

      .. note::
         When working with a text stream :py:meth:`!seek` has limitations
         beyond the general interface described here,
         which are described in more detail in :py:meth:`TextIOBase.seek`.
```

   .. method:: seek(offset, whence=SEEK_SET, /)

>       Change the stream position to the given byte *offset*,

Reviewer: It is **not** a byte offset.

I suggest also to change the name of the first argument to something other. For example to "cookie" as it is use in the code. Or to "pos" as it was in older documentation. But "offset" is misleading.

Removing references to the name further in the documentation and using "the first argument" instead of the name if it is impossible can also help, because it is hard to came with good name.

Author: The existing docs/docstrings use offset, cookie, and pos. I don't understand what's meant by cookie. I understand what's meant by offset or position. If there is no need to use the same name, there is no need for this PR.

Reviewer 2: I think *offset* is fine, as it is still an offset into the file even if not an integer byte offset. I agree we should stay away from *cookie* et al.

Reviewer: "Cookie" means an opaque value. No, it is not an offset. It is several values (position of the buffer in the file, position in the buffer, state of the encoder and decoder, packed in a 257-bit integer. See code of `_pack_cookie()` and `_unpack_cookie()` and very complex code of `tell()` and `seek()` in `_pyio.py`.

This PR is needed because semantic of `tell()` and `seek()` in binary and text files is so different.

Author: Ouch, yes TextIOWrapper is a different thing. Very unfortunate.

      and return the new absolute position as an opaque number.
      *offset* is interpreted relative to the position indicated by *whence*.
      Valid values for *whence* are:

      * :data:`SEEK_SET` -- seek from the start of the stream (the default);

>         *offset* must either be a number returned by
>         :meth:`TextIOBase.tell`, or zero.

Reviewer: It would be better to describe what will happen exactly.
```suggestion
        *offset* must either be a number returned by
        :meth:`TextIOBase.tell`, in which case the stream is returned to the position before calling :meth:`!tell()`, or zero, in which case the stream is rewound to the start.
```

        Any other *offset* value produces undefined behaviour.

>       * :data:`SEEK_CUR` -- seek from current stream position;
>         *offset* must be zero, which is a no-operation.

Reviewer: ```suggestion
      * :data:`SEEK_CUR` -- just return the current position; *offset* must be zero.
```

Author: Why?

Reviewer: Isn't it silly? "Seek, but offset must be zero, so actually do not seek." Why write seek if it does not seek? I think it only confuses users.

Reviewer 2: I suggested "seek from the current stream position; *offset* must be zero, which does not change the stream position. All other values are unsupported, and will raise an exception." above -- would this be an alright alternative?

Reviewer: Why say "seek" if it does not seek?

Author: Ok, you convinced me. How about dropping 'just':

```suggestion
      * :data:`SEEK_CUR` -- return the current position; *offset* must be zero.
```

Reviewer: Up to you. Only do not include contradictions.

Since `seek()` also returns the current position in all other cases, "return the current position" may be superfluous (it stress the difference which does not exist). Maybe just "no-operation", "does nothing" or "does not change the current position"?

Author: > Maybe just "no-operation", "does nothing" or "does not change the current position"?

Yes, I think a shorter sentence is the way to go.

>       * :data:`SEEK_SET` -- seek from the start of the stream (the default);
>         *offset* must either be a number returned by
>         :meth:`TextIOBase.tell`, or zero.
>         Any other *offset* value produces undefined behaviour.
>       * :data:`SEEK_CUR` -- seek from current stream position;
>         *offset* must be zero, which is a no-operation.
>         All other values are unsupported.
>       * :data:`SEEK_END` -- seek from the end of the stream;
>         *offset* must be zero.
>         All other values are unsupported.

Reviewer 2: ```suggestion
      * :data:`SEEK_SET` -- seek from the start of the stream (the default);
        *offset* must either be a number returned by
        :meth:`TextIOBase.tell`, or zero.
        Any other *offset* value produces undefined behaviour.
      * :data:`SEEK_CUR` -- seek from the current stream position;
        *offset* must be zero, which does not change the stream position.
        All other values are unsupported, and will raise an exception.
      * :data:`SEEK_END` -- seek from the end of the stream;
        *offset* must be zero, which does not change the stream position.
        All other values are unsupported, and will raise an exception.
```

Author: I think this is better:

```suggestion
      * :data:`SEEK_SET` -- seek from the start of the stream (the default);
        *offset* must either be a number returned by
        :meth:`TextIOBase.tell`, or zero.
        Any other *offset* value produces undefined behaviour.
      * :data:`SEEK_CUR` -- seek from the current stream position;
        *offset* must be zero, leaving the stream position unchanged.
        All other values are unsupported, and will raise an exception.
      * :data:`SEEK_END` -- seek from the end of the stream;
        *offset* must be zero, leaving the stream position at the end of the stream.
        All other values are unsupported, and will raise an exception.
```
